encoding/thrift/outbound_nowire.go

```diff
@@ -0,0 +1,248 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package thrift
+
+import (
+	"bytes"
+	"context"
+
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/protocol/stream"
+	"go.uber.org/thriftrw/wire"
+	"go.uber.org/yarpc"
+	encodingapi "go.uber.org/yarpc/api/encoding"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/thrift/internal"
+	"go.uber.org/yarpc/pkg/encoding"
+	"go.uber.org/yarpc/pkg/errors"
+	"go.uber.org/yarpc/pkg/procedure"
+)
+
+// NoWireClient is a generic Thrift client for encoding/decoding using
+// ThriftRW's "streaming" mechanisms.
+// It speaks in raw Thrift payloads.
+//
+// Users should use the client generated by the code generator rather than
+// using this directly.
+type NoWireClient interface {
+	// Call the given Thrift method.
+	Call(ctx context.Context, reqBody stream.Enveloper, resBody stream.BodyReader, opts ...yarpc.CallOption) error
+	CallOneway(ctx context.Context, reqBody stream.Enveloper, opts ...yarpc.CallOption) (transport.Ack, error)
+
+	// Enabled returns whether or not this client is enabled through a
+	// ClientOption.
+	Enabled() bool
+}
+
+// NewNoWire creates a new Thrift client.
+func NewNoWire(c Config, opts ...ClientOption) NoWireClient {
+	// Code generated for Thrift client instantiation will probably be something
+	// like this:
+	//
+	// 	func New(cc transport.ClientConfig, opts ...ClientOption) *MyServiceClient {
+	// 		c := thrift.NewNoWire(thrift.Config{
+	// 			Service: "MyService",
+	// 			ClientConfig: cc,
+	// 			Protocol: binary.Default,
+	// 		}, opts...)
+	// 		return &MyServiceClient{client: c}
+	// 	}
+	//
+	// So Config is really the internal config as far as consumers of the
+	// generated client are concerned.
+
+	var cc clientConfig
+	for _, opt := range opts {
+		opt.applyClientOption(&cc)
+	}
+
+	p := protocol.BinaryStreamer
+	if cc.Protocol != nil {
+		if val, ok := cc.Protocol.(stream.Protocol); ok {
```

Comment:
Since this check is done mainly at launch time, should we panic when a wrong protocol is given?
Also, what will happen if a stream.Protocol is given to the outbound wire implementation?

Reply:
This is almost an oversimplification - it's this way because of the current implementations.  I toyed with the idea of having a `StreamProtocol` client config option, but almost seems too much (but it does seem like the correct thing to do).